### ISSUE TITLE
.NET Core 2.1 SocketsHttpHandler proxy authentication using Windows auth is broken 

### ISSUE THREAD

**Issue by bgiot** [dotnet/corefx#30166]
When being behind a proxy requesting authentication, .net core 2.1 HttpClientHandler fails to properly authenticate against the proxy.  The following code is working in .net core 2.0 but doesn't work anymore with .net core 2.1:

```c#
            var handler = new HttpClientHandler();
            handler.DefaultProxyCredentials = CredentialCache.DefaultCredentials;

            var client = new HttpClient(handler);
            var response = client.GetAsync("https://www.mocky.io/v2/5185415ba171ea3a00704eed").GetAwaiter().GetResult();
            var content = response.Content.ReadAsStringAsync().GetAwaiter().GetResult();
```

**Comment by stephentoub** [comment]
I ran:
```C#
using System;
using System.Net;
using System.Net.Http;
using System.Threading.Tasks;

class Program
{
    static void Main()
    {
        var handler = new HttpClientHandler();
        handler.DefaultProxyCredentials = CredentialCache.DefaultCredentials;

        var client = new HttpClient(handler);
        var response = client.GetAsync("https://www.mocky.io/v2/5185415ba171ea3a00704eed").GetAwaiter().GetResult();
        var content = response.Content.ReadAsStringAsync().GetAwaiter().GetResult();

        Console.WriteLine(content);
    }
}
```
on 2.1 on Windows and I get:
```
{"hello": "world"}
```
What am I supposed to get instead?  What OS are you on?  You're using the released version of 2.1 or a preview?

**Comment by stephentoub** [comment]
> When being behind a proxy requesting authentication

Ah, I'm not behind a proxy so my test wasn't valid...

@wfurt?

**Comment by wfurt** [comment]
What platform and authentication scheme @dotconsulting ?
Best way would be positing full repro I can run.

**Comment by dotconsulting** [comment]
Windows world.  The proxy requests Windows authentication.  I don't know exactly which scheme is used. But for example, with some tools we have to define proxy settings like this : http://{domain\user}:{password}@{proxy}:{port}.  I don't know if it helps.

**Comment by karelz** [comment]
Would it help to get Fiddler/Wireshark traces? To see if we have any proxy info available at all.

**Comment by rmkerr** [comment]
Yeah, I think Fiddler or Wireshark traces would be really useful here.

**Comment by davidsh** [comment]
>Windows world. The proxy requests Windows authentication. I don't know exactly which scheme is used. But for example, with some tools we have to define proxy settings like this : http://{domain\user}:{password}@{proxy}:{port}. I don't know if it helps.

Putting the username and password in the URL is not supported.  It is considered insecure.  Our HTTP stacks in general don't support that way for both .NET Framework and .NET Core.

>handler.DefaultProxyCredentials = CredentialCache.DefaultCredentials;

The code above here assumes two things. First, that the proxy settings are found via the "default system proxy" settings. On Windows machines, that typically means IE (Internet Explorer) settings.  Second, using .DefaultCredentials only works when the proxy supports Windows authentication schemes such as Negotiate or NTLM. If the proxy only asks for Basic or Digest, then .DefaultCredentials doesn't work.  You would have to use an explicit NetworkCredential object.

**Comment by dotconsulting** [comment]
@davidsh:
Maybe back to the original message i wrote.  The code properly works with .net core 2.0 and properly authenticates against the proxy.  With .net core 2.1 we receive an "access denied" from the proxy saying no credentials are provided.  
the "username and password url" sample was just to provide some hints about how the proxy works.  
.DefaultCredentials works  with .net core 2.0 but not with .net core 2.1. I think there is a regression bug.

**Comment by davidsh** [comment]
>.DefaultCredentials works with .net core 2.0 but not with .net core 2.1. I think there is a regression bug.

In the short term, you might want to opt-out of the new SocketsHttpHandler HTTP stack in .NET Core 2.1.

See the release notes for instructions on switching back to the HTTP stack (WinHttpHandler for Windows) that was used in .NET Core 2.0:

https://github.com/dotnet/core/blob/master/release-notes/2.1/2.1.0.md

Adding a line of code like this to your app will switch you back to the .NET Core 2.0 HTTP stack:
>AppContext.SetSwitch("System.Net.Http.UseSocketsHttpHandler", false);

**Comment by dotconsulting** [comment]
I'll try that workaround.  I'll keep out posted. Thanks!

**Comment by rmkerr** [comment]
This isn't a case I considered when writing the default proxy string parser, so if WinHttp supports providing credentials that way there is a regression here.

**Comment by dotconsulting** [comment]
@rmkerr:
The issue is with .DefaultProxyCredentials = CredentialCache.DefaultCredentials;
It seems that the credentials (the current logged user) are not passed to the proxy for authentication.

**Comment by davidsh** [comment]
@dotconsulting 

>@rmkerr:
The issue is with .DefaultProxyCredentials = CredentialCache.DefaultCredentials;
It seems that the credentials (the current logged user) are not passed to the proxy for authentication.

Can you try an experiment?  Instead of using `CredentialCache.DefaultCredentials`, can you use something like this `new NetworkCredential("username", "password", "domain");` ?

Just want to see if the problem is .DefaultCredentials or a generalized problem with Windows Authentication for proxy handling in SocketsHttpHandler?  Thx.

**Comment by dotconsulting** [comment]
@davidsh 
I already tried what you proposed.  Same issue when using a specific NetworkCredential instance.  I tried also to play with Proxy parameter with a WebProxy class but not better.

**Comment by davidsh** [comment]
>I already tried what you proposed. Same issue when using a specific NetworkCredential instance. I tried also to play with Proxy parameter with a WebProxy class but not better.

Thanks for trying.  This looks like there is a bug with Windows Authentication for proxy credentials in SocketsHttpHandler. 

To speed up our investigation, is it possible for you to take WireShark traces?  I'd like to see if there is actual network traffic attempting Windows authentication (but failing) or just not attempting it at all.  Please remove any confidential information from the traces.  Thx.

**Comment by dotconsulting** [comment]
@davidsh 
Sure I'll try to produce a WireShark trace.  I'm in Belgium, so now it's evening time (8:00pm).  I'll be at the office tomorrow. Some advice to setup WireShark with .net core console ?
Thanks!

**Comment by davidsh** [comment]
>Some advice to setup WireShark with .net core console ?

There is nothing specific w/ .NET Core regarding WireShark.  Once you set it up on your client machine, you can start the trace, run your app long enough to produce the failure and then stop tracing.  Then you can process the trace and remove unneeded TCP sessions, etc. from it to reduce it down.  

If you haven't used WireShark before, then you can go here to get more info:
https://www.wireshark.org/

**Comment by davidsh** [comment]
I've reproduced the bug.  SocketsHttpHandler isn't even beginning the Windows authentication exchange.

I set up a repro using a test proxy server that uses Negotiate (it actually isn't a fully working implementation of Negotiate..but it's enough to see there is a problem with SocketsHttpHandler).
 
On .NET Core 2.1 using SocketsHttpHandler.  I see this exchange thru the proxy:

```
W:\HappySockets>happysockets proxy server -Negotiate
Listening on port 3128 using authentication Negotiate ...
Connected: [::1]:1436
GET http://corefx-net.cloudapp.net/echo.ashx HTTP/1.1
Host: corefx-net.cloudapp.net

GET /echo.ashx HTTP/1.1
Host: corefx-net.cloudapp.net
X-Forwarded-For: ::1


HTTP/1.0 407 Proxy Authentication Required
Proxy-agent: Netscape-Proxy/1.1
Proxy-Authenticate: Negotiate
Content-Length: 0
Connection: close
Date: Wed, 06 Jun 2018 19:06:24 GMT
```

So, you can see that SocketsHttpHandler didn't even try to send Windows credentials.

On .NET Core 2.1, using the AppContextSwitch to switch back to WinHttpHandler, I see this (note: I edited the Negotiate strings to remove confidential info).  Also note that WinHttpHandler (which uses WinHTTP) is a little inefficient here since it starts a new TCP connection to begin the Negotiate.

```
W:\HappySockets>happysockets proxy server -Negotiate
Listening on port 3128 using authentication Negotiate ...
Connected: [::1]:1633
GET http://corefx-net.cloudapp.net/echo.ashx HTTP/1.1
Proxy-Connection: Keep-Alive
Accept-Encoding: peerdist
X-P2P-PeerDist: Version=1.1
X-P2P-PeerDistEx: MinContentInformation=1.0, MaxContentInformation=2.0
Host: corefx-net.cloudapp.net

GET /echo.ashx HTTP/1.1
Proxy-Connection: Keep-Alive
Accept-Encoding: peerdist
X-P2P-PeerDist: Version=1.1
X-P2P-PeerDistEx: MinContentInformation=1.0, MaxContentInformation=2.0
Host: corefx-net.cloudapp.net
X-Forwarded-For: ::1

HTTP/1.0 407 Proxy Authentication Required
Proxy-agent: Netscape-Proxy/1.1
Proxy-Authenticate: Negotiate
Content-Length: 0
Connection: close
Date: Wed, 06 Jun 2018 19:08:39 GMT

Connected: [::1]:1634
GET http://corefx-net.cloudapp.net/echo.ashx HTTP/1.1
Proxy-Connection: Keep-Alive
Accept-Encoding: peerdist
X-P2P-PeerDist: Version=1.1
X-P2P-PeerDistEx: MinContentInformation=1.0, MaxContentInformation=2.0
Host: corefx-net.cloudapp.net
Proxy-Authorization: Negotiate YH4GBisGAQUFA...

GET /echo.ashx HTTP/1.1
Proxy-Connection: Keep-Alive
Accept-Encoding: peerdist
X-P2P-PeerDist: Version=1.1
X-P2P-PeerDistEx: MinContentInformation=1.0, MaxContentInformation=2.0
Host: corefx-net.cloudapp.net
Proxy-Authorization: Negotiate YH4GBisG...
X-Forwarded-For: ::1


HTTP/1.0 407 Proxy Authentication Required
Proxy-agent: Netscape-Proxy/1.1
Proxy-Authenticate: Negotiate
Content-Length: 0
Connection: close
Date: Wed, 06 Jun 2018 19:08:39 GMT
```

**Comment by davidsh** [comment]
It seems that SocketsHttpHandler is requiring the proxy to send back this header in order to do any Windows authentication:

```
Proxy-Support: Session-Based-Authentication
```

This header normally describes how a proxy supports connection persistence to a destination server.  That persistence is required in order to do Windows auth schemes such as Negotiate/NTLM to a destination server (i.e. 401).  However, this header doesn't necessiarly describes client requirements when doing Windows auth to a proxy server (i.e. 407).

When I modified my test proxy server to send back the "Proxy-Support: Session-Based-Authentication" response header, I got a different behavior from SocketsHttpHandler.  It threw an exception:

>System.Net.Http.HttpRequestException: Authentication failed because the connection could not be reused.
   at System.Net.Http.HttpConnection.DrainResponseAsync(HttpResponseMessage response)
   at System.Net.Http.AuthenticationHelper.SendWithNtAuthAsync(HttpRequestMessage request, Uri authUri, ICredentials credentials, Boolean isProxyAuth, HttpConnection connection, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.AuthenticationHelper.SendWithAuthAsync(HttpRequestMessage request, Uri authUri, ICredentials credentials, Boolean preAuthenticate, Boolean isProxyAuth, Boolean doRequestAuth, HttpConnectionPool pool, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.DiagnosticsHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at NetCoreProxyTest.Program.Main(String[] args) in C:\Users\dshulman\Documents\Visual Studio 2017\Projects\NetCoreProxyTest\NetCoreProxyTest\Program.cs:line 27

So, there is probably a few different things to fix in SocketsHttpHandler.

**Comment by bgiot** [comment]
I tried to use AppContext.SetSwitch("System.Net.Http.UseSocketsHttpHandler", false);

But then I got an exception coming from WinHttpAuthHelper class:

System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Net.Http.WinHttpAuthHelper.CheckResponseForAuthentication(WinHttpRequestState state, UInt32& proxyAuthScheme, UInt32& serverAuthScheme)
   at System.Net.Http.WinHttpHandler.StartRequest(WinHttpRequestState state)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at TestProxyAuthentication.Program.Main(String[] args) in C:\Users\giotb\source\temp\TestProxyAuthentication\TestProxyAuthentication\Program.cs:line 21

**Comment by DaniilSokolyuk** [comment]
Nuget affected too
https://github.com/NuGet/Home/issues/6978

**Comment by davidsh** [comment]
@dyamo 

>It doesn't work in 2.1, regardless of whether 'System.Net.Http.UseSocketsHttpHandler' is true or false, although an exception is thrown if it is false.

This is suprising.  When the AppContext switch is used to switch off SocketsHttpHandler, proxy detection  and use is done by the original WinHttpHandler HTTP stack.  I haven't seen any problems with that.

Can you show a repro and the callstack / exception you're getting?  Thx.

**Comment by davidsh** [comment]
@bgiot 

>But then I got an exception coming from WinHttpAuthHelper class:
System.NullReferenceException: Object reference not set to an instance of an object.
at System.Net.Http.WinHttpAuthHelper.CheckResponseForAuthentication(WinHttpRequestState state, UInt32& proxyAuthScheme, UInt32& serverAuthScheme)
at System.Net.Http.WinHttpHandler.StartRequest(WinHttpRequestState state)
at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
at TestProxyAuthentication.Program.Main(String[] args) in C:\Users\giotb\source\temp\TestProxyAuthentication\TestProxyAuthentication\Program.cs:line 21

Can you describe in more detail how the proxy is configured in your scenario?  What are the IE settings?  Is it auto-discovery...i.e. looking for a PAC file on the network?  What kind of proxy server is it?  Etc.  Thx.

**Comment by bgiot** [comment]
@davidsh 
"Automatically detect settings" is checked in the IE settings.  So, probably a pac file loaded from the network.

**Comment by davidsh** [comment]
What about the kind of proxy server?  Is it using Negotiate or NTLM (or both)?  Can you do some WireShark traces of this?  We'd like to understand the error you're getting when using WinHttpHandler.  There are probably 2 different bugs here: one with SocketsHttpHandler that we understand already, second with WinHttpHandler that we can't repro yet.

**Comment by davidsh** [comment]
@bgiot 

I was able to reproduce this exception in WinHttpHandler when I use a PAC file.  I used explicit PAC file settings instead of "automatic discovery" since it was easier to repro it that way.  But I got the same NullReferenceException.

![image](https://user-images.githubusercontent.com/7684835/41113984-f179e8ac-6a37-11e8-8d35-022c55383431.png)

**Comment by davidsh** [comment]
We will probably split this bug into separate bugs since the root causes are different between WinHttpHandler vs. SocketsHttpHandler.

**Comment by rmkerr** [comment]
I think it's likely that the WinHttpHandler regression is related to [this change](https://github.com/dotnet/corefx/commit/6d031a04f3df8fadd5b82d9c23d904182eea94e4) that I made. It changed the code where you're seeing the NRE.

**Comment by davidsh** [comment]
>I think it's likely that the WinHttpHandler regression is related to this change that I made. It changed the code where you're seeing the NRE.

Yes.  I understand the fix but it will have some restrictions to it because of #6997.

**Comment by rmkerr** [comment]
Oh interesting. Thanks for pointing that out, I wasn't aware of those restrictions.

**Comment by tebeco** [comment]
Anyone here knows how to create :
* 1 docker network

2 image on that network
* 1 docker image running as a server + NTLM Proxy
* 1 docker image (windows container) with the previous NTLM proxy as proxy in the IE settings

it will help for repro here and there
and it should also be part of non regression test for CoreFx stuff around proxy / socket etc ... it would avoid surprise i major release ^^

**Comment by davidsh** [comment]
>and it should also be part of non regression test for CoreFx stuff around proxy / socket etc ... it would avoid surprise i major release ^^

FYI, we are in the process of working on an Enterprise-Scenario capability for CoreFx testing.

**Comment by davidsh** [comment]
@bgiot 

>"Automatically detect settings" is checked in the IE settings. So, probably a pac file loaded from the network. (Please note - in case of - I'm the creator of this issue entry. I just changed my profile from dotconsulting to bgiot; thanks!)

I was having some trouble getting a repro for this.  Can you verify a few things for me with your repro case? 

You indicated that your IE settings was "Automatically detect settings".  Is that correct?

And you were using HttpClient similar to this?

```c#
var handler = new HttpClientHandler();
handler.DefaultProxyCredentials = CredentialCache.DefaultCredentials;
var client = new HttpClient(handler);
HttpResponseMessage response = await client.GetAsync(url);
```

Are you getting back a 'response' with status code of '407'?

**Comment by tebeco** [comment]
> FYI, we are in the process of working on an Enterprise-Scenario capability for CoreFx testing.

i’m curious since i did not take a look at all the repo / tests details, how could these proxy test could have passed

**Comment by davidsh** [comment]
@bgiot 

Is it possible for you to take traces of your application that is failing with '407 status code?

Please follow the instructions here:
https://github.com/dotnet/corefx/blob/master/Documentation/debugging/windows-instructions.md

And follow the "Using PerView" instructions.  For the additional providers, please list at least "*Microsoft-System-Net_http".  


We're trying to reproduce what you're seeing.  We do know that there is a bug in SocketsHttpHandler when using Windows Authentication for Proxy auth and the Proxy closes the connection (as your repro scenario shows):

```
Connection: close
```

But that bug would show up as an HttpRequestException and not give you a 407 status code. So, we'd like more information to understand your scenario and enterprise network topology.

**Comment by davidsh** [comment]
@bgiot Thanks for the traces!  They were very helpful!

I was able to reproduce the problem now.  The traces helped remind me that you are connecting to an HTTPS endpoint thru the authenticating proxy.  I was originally trying to repro this with an HTTP endpoint thru the proxy.

**Comment by davidsh** [comment]
@bgiot #30330 is now tracking your bug since it is specific to HTTPS endpoints going thru an authenticating proxy that uses Windows authentication.  So, please track that issue.

**Comment by tebeco** [comment]
how did we end up with surprise like this ?
i mean entreprise context and proxy is kind frequent no ?

**Comment by karelz** [comment]
@tebeco proxy support in SocketsHttpHandler came in quite late (to support also enterprise scenarios). We tested NTLM and Basic, but we couldn't get easily to Kerberos proxy authentication setup. That's why it slipped through.
Automated testing of these setups are fairly involved and costly. It is a known test gap we have :( ... we started investing into it quite recently, so I hope we will be soon in position to protect against regressions in the space.

**Comment by sabiland** [comment]
I have experienced similar issues with HttpClient after we upgraded to ASP Net Core 2.1. 
Bu only with few (1 or 2) calls to some of our company-internal API's.

This:

`AppContext.SetSwitch("System.Net.Http.UseSocketsHttpHandler", false);`

solved our issues. Though, is this a **temporary** quick-solution?

**Comment by karelz** [comment]
@sabiland your "solution" will fall back to old networking stack. It is a reasonable temporary workaround to get unblocked until we release fix for 2.1.
In the meantime, can you please confirm that your failing cases are indeed duplicates of what @davidsh identified? (see instructions in https://github.com/dotnet/corefx/issues/30166#issuecomment-396326804 or you can capture network traffic via Wireshark/Fiddler)

**Comment by sabiland** [comment]
Ok, I understand it is a temporary solution.

For now I can say only, that the problem was only when calling **one** (among many others) in-house API. The result was always timeout ("...Task was cancelled..."). So Exception message didn't provide any extra info.

How could I check what exactly was the problem? OK, I'll check #30166.

**Comment by karelz** [comment]
Even if it was just 1 endpoint / in-house API, please follow the instructions I linked above. Let's make sure we understand the real root-cause.

**Comment by csharmath** [comment]
This causes PowerShell Core 6.1 to fail with SPNego as well, so it's not just with Proxies but seems to be an issue with Negotiate auth in general.

i.e.: SocketsHttpHandler not using Negotiate / SPNego

This PowerShell issue has some more details as well.
https://github.com/PowerShell/PowerShell/issues/7801

Testing
----
As per testing this, IIS can be configured to only support Negotiate:Kerberos as the windows auth provider. i.e.: disable NTLM. So maybe you can do a quicker / easier / cheaper setup with IIS to make sure that SocketsHttpHandler works with negotiate.

@PatrickLang has a work in progress for how to setup containers with gMSA's SPN's, etc for a Kerberos setup https://github.com/MicrosoftDocs/Virtualization-Documentation/pull/592

**Comment by karelz** [comment]
@csharmath I am not 100% sure how is your comment related to this issue.
Given that the main investigation with proxies has been moved into separate issues #30327, #30330, and both have been addressed in 2.1.x, I would recommend to:
* Try your repro on 2.1.x (daily builds as the last fix went into not yet-released 2.1.5 release).
* If it still repros, isolate a repro without PowerShell (as part of minimizing the problem) and file a new bug.

**Comment by tebeco** [comment]
i think the biggest issue is that now `SocketsHttpHandler` supports the Negociate it still does not goes for `PlatformSpecific` api to get the OS value

in Corporation don’t ever try to use env var like `HTTP_PROXY` (even with `NO_PROXY`)

for example my current client have 3 proxy url and use a `PAC` script to resolve the proxy url to use based on the url we try to reach

i would expect some code to check the current Platform and use respective standard in each one of them to actually get that last piece of code working

every single third part on nuget using HttpClieny is broken right now if they did not designed the API so you can pass the proxy url (wich is bad for PAC scenario)

**Comment by karelz** [comment]
@tebeco good points (esp. on the `dotnet restore` scenario). Let's take the discussion into separate issue. We'll be interested in concrete information on which parts of proxy don't work today on which OS.
To my best knowledge we addressed all known problems and differences from .NET Core 2.0 in 2.1.5 (to be released). If you are aware of scenarios which still don't work (with 2.1.5 daily builds), please file them as issues, so that we can address them. Thanks!

As I mentioned earlier in https://github.com/dotnet/corefx/issues/30166#issuecomment-396725994, the proxy support came in late. We had the option to not support proxies at all in 2.1, or to make an honest attempt with potential tail of bugs. While it is not ideal, it went quite ok IMO - we had a few bugs reported, couple of dozens customers impacted, but it was not super-massive + there is always opt-out to the old way.

**Comment by tebeco** [comment]
> To my best knowledge we addressed all known problems and differences from .NET Core 2.0 in 2.1.5 (to be released)

you lost me here, 2.1.5 runtime (hence the TBR ?) ?

> there is always opt-out to the old way.
you mean the env var to disable `SocketHttpHandler` ? i think i tried it already and did not seem to affect the restore

not sure what i missed

**Comment by karelz** [comment]
> you lost me here, 2.1.5 runtime (hence the TBR ?) ?
> something is this release should affect the http handlers ?

AFAIK all known proxy related bugs are fixed in 2.1.5 (to be released).

> you mean the env var to disable `SocketHttpHandler`? i think i tried it already and did not seem to affect the restore

Yes, that is what I meant. Maybe dotnet/nuget are doing something special and use `SocketsHttpHandler` directly. That would be specific to them - please file a new bug (in CLI repo I think) and let's get to the bottom of this there.

**Comment by tebeco** [comment]
i’m not sure moving to the CLI is a good idea
i feel like it should be dealt here

for example when using `AddAzureKeyVault` on aspnet core bootstrap i have to implement my own `IKeyVaultClient`

it could break CosmosDb api (i did not check if it does not allows to feed an httpClient)

same question for EfCore

**Comment by tebeco** [comment]
i just noticed that you saw this one
https://github.com/dotnet/corefx/issues/24574

i could try 2.1.5 but honestly the main issue is the loss of windows specific code
so unless all Platform align on a real standard (and i mean a standard used by corporation)
if not, coreFx will have to use Platefrom specific api
something like the Windows Compatibility pack :

the httpclient is instanciated using reflection to check if it’s running using fullfx or netcore to call a `private/internal` ctor. Why not resolving the proxy at that time too ?

**Comment by csharmath** [comment]
@karelz thanks for pointing out that I am mixing a different topic to this proxy only issue, adding a new issue about SPNego.

**Comment by tebeco** [comment]
@karelz so yesterday we tried to use CosmosDb with DocumentDb client api
i’m not sure we found the actual repo for DocumentDb.Core (the netstandard package)
still after décompilation it seems that a private handler is is nested next to the httpclient and all the API is closed

if you look for “DocumentDb” you’ll find answers telling you to do :
```
<system.net>
    <defaultProxy enabled="true" useDefaultCredentials="true">
        <proxy usesystemdefault="True" />
    </defaultProxy>
</system.net>
```

this is why this issue probably need to be dealt in coreFx and not over every single repo
i’m surprised there’s not much people around asking why lots of there third part tool stopped working when switching to netcoreapp ^^
* either corporation don’t bother anymore with proxy (i highly doubt it)
* did they rolled back to fullFx
* did they used tools like CNTLM (wich is really bad)
* only few moved to netcoreapp (weird because i would say no)

so i wonder as there was on Opt_out from SocketHandler
if there no way to add an opt_in only to inject the proxy from si does settings in anyway

**Comment by tebeco** [comment]
@karelz so yesterday we tried to use CosmosDb with DocumentDb client api
i’m not sure we found the actual repo for DocumentDb.Core (the netstandard package)
still after décompilation it seems that a private handler is is nested next to the httpclient and all the API is closed

if you look for “DocumentDb” you’ll find answers telling you to do :
```
<system.net>
    <defaultProxy enabled="true" useDefaultCredentials="true">
        <proxy usesystemdefault="True" />
    </defaultProxy>
</system.net>
```

this is why this issue probably need to be dealt in coreFx and not over every single repo
i’m surprised there’s not much people around asking why lots of there third part tool stopped working when switching to netcoreapp ^^
* either corporation don’t bother anymore with proxy (i highly doubt it)
* did they rolled back to fullFx
* did they used tools like CNTLM (wich is really bad)
* only few moved to netcoreapp (weird because i would say no)

so i wonder as there was on Opt_out from SocketHandler
if there no way to add an opt_in only to inject the proxy from si does settings in any way
* Context switch ?
* platform specific code using compat pack code ?
* a Configuration Section (kinda like the fullfx way)

again ... we cannot control third part that were released and was never designed to take care of proxy properly

**Comment by karelz** [comment]
@tebeco it seems that we have duplicate conversation here and in #24574. It is also confusing to see several scenarios thrown into the same pile and mixes.
Let's try to untangle things:

The main problem seems to be that proxy on Windows does not use default credentials unless you explicitly ask for it via `HttpClientHandler.DefaultProxyCredentials = CredentialCache.DefaultCredentials`. That can be problematic in 3rd party libraries if they do not expose HttpClientHandler configuration.
The scenario seems to work with disabled `SocketsHttpHandler` on Windows. It also works fine on .NET Framework if you set config file to `<defaultProxy enabled="true" useDefaultCredentials="true">`.
The ask is to provide a way (default or explicit) to set process-wide proxy credentials (or even proxy settings), without need to pass the information programatically to each HttpClientHandler created in the process (incl. those created by 3rd party libraries).

Current workaround: Disable SocketsHttpHandler in the app via env. settings.

Is that good summary?
If yes, we should file a new issue to clearly track the feature ask. It seems legit to me and hinted in other issues as well. It could use good title and clear description though.

Did I miss other problems / bugs / suggestions in your replies?

**Comment by tebeco** [comment]
i will check again tomorrow (from Paris here)
as i understand from another MSFT voice

Msft has no purpose on even trying to make proxy detect IE settings 

they try to convince corporation to use transparent proxy unless they (corpo) prouve msft that they (msft) will loose money

also i think i did try to opt out from `sockethandler` with no luck 

does it work for windows service without `app.config` ?

**Comment by karelz** [comment]
@tebeco I don't quite understand what you meant in your last post. Can you please rephrase /explain in details what you meant?

> make proxy detect IE settings

Currently we use IE settings. Why do you think we don't?

> they try to convince corporation to use transparent proxy unless they (corpo) prouve msft that they (msft) will loose money <== can you confirm / inform that ?

No idea what you meant here. I would suggest to stick to facts - what works vs. what does not. Let's avoid any speculations what MS is trying to convince developers to do.
I can speak on behalf of Networking area - believe me, we are just trying to light up all important scenarios (in priority order) and we are trying to understand which scenarios we missed. BTW: You're the first one to poke into this scenario.

> also i think i did try to opt out from `sockethandler` with no luck

Can you please double check? If it never worked even in .NET Core 2.0 (same as disabled SocketsHttpHandler), that would be good to know.

> does it work for windows service without app.config?

What should work where without app.config? On .NET Framework I assume app.config is necessary.

**Comment by tebeco** [comment]
i will when i’ll be sober tomorrow (and i will delete this one)

**Comment by tebeco** [comment]
@karelz done, sry about that

**Comment by JohannesHoppe** [comment]
> 3rd party libraries 

This also includes the Azure WebJobs SDK. I was never able to get it running behind our evil company proxy ('System.Net.Http.UseSocketsHttpHandler' does not work). For several years .Net Core is an endless source of frustration. 👎

**Comment by tolgabalci** [comment]
In case anyone else runs in to this, using HttpClient to call out to WEBAPIs inside of IIS was causing the error:

`HttpRequestException: Response status code does not indicate success: 403 (This URL is not allowed to be proxied)`

When troubleshooting with Fiddler, the error would go away and everything would work fine.

Solution was to add the following to startup.cs:
`AppContext.SetSwitch("System.Net.Http.UseSocketsHttpHandler", false);`

**Comment by davidshen84** [comment]
dotnet --version
    2.1.700

I am still seeing this issue, and my company does not have a proxy. Any update on the status of this bug?

**Comment by davidsh** [comment]
>I am still seeing this issue, and my company does not have a proxy. Any update on the status of this bug?

This issue is closed.  It was fixed in .NET Core 3.0 which is available in Preview right now.

**Comment by vologar** [comment]
Same issue here. Trying to connect from Azure function emulator to Azure blob storage. TCP Viewer shows the calls are attempted directly to Azure ip without any knowledge of Proxy settings. Net version woks as expected. Setting AppContext.SetSwitch("System.Net.Http.UseSocketsHttpHandler", false) resolved the issue, but we had to spend hours to understand the root course. Has it ever been tested from behind the firewalls ?

**Comment by vologar** [comment]
Azure Functions SDK  still uses .Net Core 2.*

**Comment by davidsh** [comment]
This issue was only fixed in .NET Core 3.0. It was not fixed in .NET Core 2.*. That's why you still see the problem.

When Azure Functions SDK is upgraded to .NET Core 3.x, then the problem should be resolved for you.
